include/private/entry.h

```diff
@@ -27,24 +27,137 @@
+ * and adding it to the severity value which is according to the 
+ * RFC 5424 Section 6.2.1. The shift operation is done prior to get_prival()
+ * function with macros in "facility.h"
+ *
+ * **Thread Safety: MT-Safe**
+ * This function is thread safe.
+ *
+ * **Async Signal Safety: AS-Safe **
+ * This function must be safe to call from signal handlers
+ *
+ * **Async Cancel Safety: AC-Safe**
+ * This function must be safe to call from threads that may be asynchronously
+ * cancelled.
+ *
+ * @since release v2.0.0
+ *
+ * @param facility Facility value. This should be a \c STUMPLESS_FACILITY value.
+ * 
+ * @param severity Severity value. This should be a \c STUMPLESS_SEVERITY value
+ *
+ * @return Priority value
+ */
 int
 get_prival( enum stumpless_facility facility,
             enum stumpless_severity severity );
 
+/**
+ * Locks the mutex within the entry.
+ *
+ * **Thread Safety: MT-Safe**
+ * This function is thread safe.
+ *
+ * **Async Signal Safety: AS-Unsafe **
+ * This function is not safe to call from signal handlers due to the use of
+ * function 'pthread_mutex_lock()' from <pthread.h> that is not guarenteed to 
+ * be async signal safe.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * This function is not safe to call from threads that may be asynchronously
+ * cancelled, due to the use of lock that could be left in undefined state. 
+ *
+ * @since release v2.0.0
+ *
+ * @param entry The entry to be locked.
+ *
+ */
 void
 lock_entry( const struct stumpless_entry *entry );
 
+/**
+ * Adds the element to the entry.
+ *
+ * **Thread Safety: MT-Unsafe**
+ * This function is not thread safe as it doesn't use any synchronization 
+ * or locking mechanisms while accessing shared resources.
+ *
+ * **Async Signal Safety: AS-Unsafe**
+ * This function is not safe to call from signal handlers due to the use of
+ * memory functions.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * This function is not safe to call from threads that may be asynchronously
+ * cancelled, due to the use of memory functions that could leave the 
+ * memory in undefined state.
+ *
+ * @since release v2.6.0
+ *
+ * @param entry The entry to add the new element.
+ *
+ * @param element The new element to add to entry.
+ *
+ * @return The modified entry if no error is encountered. If an error is
+ * encountered NULL is returned.
+ */
 struct stumpless_entry *
 locked_add_element( struct stumpless_entry *entry,
                     struct stumpless_element *element );
 
+/**
+ * Returns the element located at index within the entry.
+ *
+ * **Thread Safety: MT-Unsafe**
+ * This function is not thread safe as it doesn't use any synchronization 
+ * or locking mechanisms while accessing shared resources.
+ *
+ * **Async Signal Safety: AS-Safe**
+ * This function is safe to call from signal handlers.
+ *
+ * **Async Cancel Safety: AC-Safe**
+ * This function is safe to call from threads that may be asynchronously 
+ * canceled.
```

Comment:
This function is unfortunately AS and AC-unsafe for the same reasons listed in `include/private/error.h` for `raise_error`.

There is some nuance to this and possibly the reasons in on `raise_error` could be updated, and perhaps a case made for `AC-Safe` currently, but for now you'll need to match the `raise_error` safety to be consistent.

Reply:
Hmm this is because this uses raise_index_out_of_bounds() right? are there other things too btw? since i didn't checked for async safety before i sometimes doubt about what to write.

Reply:
That's correct, it's because of the index error, which in turn calls the generic error raising function. I'll need to make the error handling safer at some point, but for now it's marked as unsafe for signals or async cancellation.

If you're asking whether I reviewed your safety determinations for the other functions as well, I did and they looked right to me!

Reply:
Thanks for the feedback i appreciate it :)